Draft: content and planning of building and packaging chapter (#3)
Simulation-Software-Engineering/Lecture-Material: building-and-packaging/material/packaging_python_slides.md

---
type: slide
slideOptions:
  transition: slide
  width: 1400
  height: 900
  margin: 0.1
---

<style>
  .reveal strong {
    font-weight: bold;
    color: orange;
  }
  .reveal p {
    text-align: left;
  }
  .reveal section h1 {
    color: orange;
  }
  .reveal section h2 {
    color: orange;
  }
</style>

# How to prepare a Python code for packaging

---

## What is PyPI?

- [PyPI](https://pypi.org/) = **The Python Package Index** is a repository of software developed in the Python community
- PyPI itself is developed on GitHub through another software called [Warehouse](https://github.com/pypa/warehouse)

> - PyPI has a informative [public dashboard](https://p.datadoghq.com/sb/7dc8b3250-85dcf667bd?from_ts=1636972832484&to_ts=1636976432484&live=true) to show its activity

Reviewer: ```suggestion
- PyPI has an informative [public dashboard](https://p.datadoghq.com/sb/7dc8b3250-85dcf667bd?from_ts=1636972832484&to_ts=1636976432484&live=true) to show its activity
```

- Who maintains PyPI and pip? A big community! For example check out [Donald Stufft](https://twitter.com/dstufft/)

> - A major advantage is the active maintenance of PyPI, which is fueled by a big community. PyPI is also on [Twitter](https://twitter.com/PyPI)

Reviewer: ```suggestion
- A major advantage is the active maintenance of PyPI, which is fueled by a big community.
```
Who is not? :D 
But if you want to keep it, keep it.

---

## Example of PyPI package: fenicsprecice

- Having a look at [fenicsprecice](https://pypi.org/project/fenicsprecice/)

---

## What is pip?

- pip is sometimes referred to as a *package installer* and sometimes as a *package management system*

> - For all practical purposes it is an installer for packages available at PyPI
> - pip is itself a [package](https://pypi.org/project/pip/) which is available on PyPI

Reviewer: ```suggestion
- For all practical purposes, it is an installer for packages available at PyPI
- pip is itself a [package](https://pypi.org/project/pip/), which is available on PyPI
```
on or at PyPI? :) I don't know what the best answer is, but consistency prevents you from students starting to think about it here :wink:

- pip is open-source and is developed on [GitHub](https://github.com/pypa/pip)
- pip has many more things which are worth exploring! [Look here!](https://pip.pypa.io/en/latest/development/)

---

> ## Python Enhancement Proposals (PEPs)

Reviewer: I could not follow the connection from PyPI / pip to PEPs

Author: What I intended to convey was that most of the standardizations around pip and PyPI are in fact defined in some PEP. They dont become standardization's just like that, but the PyPA accepts them as a PEP

- PEP is an evolving deisgn document which provides information regarding new features of Python, new processes and new environments

> - PEPs typically involve concise technical information which also acts as standardizations

Reviewer: ```suggestion
- PEPs typically involve concise technical information, which also acts as standardizations
```

> - Packaging workflows are also standardized in through multiple PEPs. Examples are:

Reviewer: ```suggestion
- Packaging workflows are also standardized through PEPs. Example:
```

    - [PEP 427](https://www.python.org/dev/peps/pep-0427/) introduces the built-package format "wheel"
- It is always a good idea to look for **PEP** conventions before trying out processes

---

## Questions?

---

## File structure for code packaging 1/2

- Most essential component of packaging is file structure. Basic file structure is as follows:

```bash
package_name/
└── src/
    └── solver/
        ├── __init__.py
        └── solver.py
```

- The file `__init__.py` is required to import the `src/` as a package. This file is mostly empty

> - `solver.py` will contain the code developed by the user

Author: I see the mistake more as having a redundant `src/` layer. A not-so-hard nomenclature is to have the source folder name the same as the main solver file name. This is what we follow in `fenicsprecice`, we have `fenicsprecice/fenicsprecice.py` there.

> - `solver.py` will contain the code developed by the user

Reviewer: But solver is here really just an example, right? That was not clear to me when reading. What would be a good generic name?

> - `solver.py` will contain the code developed by the user

Reviewer 2: I would have said something like `modulename` would be a generic name. At least with the structure in the slides one uses this solver package as `from solver import solver`. I guess when one uses `modulename` for the directory one would use `from modulename import solver`. In my opinion the latter way of importing the package/module is nicer since it also makes clear where the file and where the directory name play a role.

> - `solver.py` will contain the code developed by the user

Reviewer 2: Maybe it should rather be

```text
repository_name/
└── src/
    └── package_name/
        ├── __init__.py
        └── solver.py
```
to avoid confusstion about the package and module names? I assume a package could have several modules.

> - `solver.py` will contain the code developed by the user

Reviewer 2: The setuptools documentation even shows [a nice example](https://setuptools.pypa.io/en/latest/userguide/quickstart.html) that rather lookes like
```
package_name/
└── src/
    └── package_name/
        ├── __init__.py
        └── solver.py
```

The setuptools example also uses a `setup.cfg` and avoids the `setup.py`. Is the `setup.py` the old style that should be deprecated in the future? If so, it would maybe make sense to set up the installation process once with a `setup.py` and once with a `setup.cfg` in the exercise so one know both ways, but especially knows the new (=good/better) way of defining the installation procedure.

> - `solver.py` will contain the code developed by the user

Author: `solver/` is just an example in this case yes. `modulename` is more generic than solver for sure.

> - `solver.py` will contain the code developed by the user

Reviewer 2: Isn't it bad practice that your package and module have the same name? I find `from solver import solver` ambiguous. It also confused me a bit when doing the exercise. Why not use the term `package_name` for the directory? It makes the whole slide more verbose to me. I see which part of the directory structure defines the package name. 

Using the current naming: If I just write `import solver` would it import everything inside `solver/` or does it import `solver/solver.py`? Maybe it is clear to someone who is an expert to Python, but to me it is not immediately clear.

> - `solver.py` will contain the code developed by the user

Author: I am going with

```bash
package_name/
└── src/
    └── package_name/
        ├── __init__.py
        └── solver.py
```

│   └── solver/
│       ├── __init__.py
│       └── solver.py
└── tests/
```

- We will now look at individual components in greater detail

---

## setup.cfg

- We will use [setuptools](https://pypi.org/project/setuptools/) to define the installation recipe for the package
- `setuptools` can be configured using a config script called `setup.cfg`
- `setup.cfg` consists of details like name of the package, version of package, files to include in the package and more.
- An example of `setup.cfg` is:

```bash
[metadata]
name = example-pkg-YOUR-USERNAME-HERE
version = 0.0.1
author = <name>
author_email = <email>
description = <Add your description here>
long_description = file: README.md
long_description_content_type = text/markdown
url = https://github.com/Simulation-Software-Engineering/test-exercise-packaging
project_urls =
    Bug Tracker = https://github.com/Simulation-Software-Engineering/test-exercise-packaging/issues
classifiers =
    Programming Language :: Python :: 3
    License :: OSI Approved :: <Name of license>
    Operating System :: OS Independent

[options]
package_dir =
    = solver
packages = find:
python_requires = >=3.6

[options.packages.find]
where = solver

> ```

Author: I plan to show this on the slide and then show the corresponding file in [fenicsprecice](https://github.com/precice/fenics-adapter). I now realize that this is perhaps too big for one slide, I will cut the code block to have bare minimum information

- An example of `setup.py` is:

```bash
import setuptools

with open("README.md", "r", encoding="utf-8") as fh:
    long_description = fh.read()

setuptools.setup(
    name="YOUR-USERNAME-SSE-package",
    version="0.0.1",
    author="Your Name",
    author_email="Your Email",
    description="A small description",
    long_description=long_description,
    long_description_content_type="text/markdown",
    url="https://github.com/Simulation-Software-Engineering/test-exercise-packaging",
    project_urls={
        "Bug Tracker": "https://github.com/Simulation-Software-Engineering/test-exercise-packaging/issues",
    },
    classifiers=[
        "Programming Language :: Python :: 3",
        "License :: OSI Approved :: <Name of License>",
        "Operating System :: OS Independent",
    ],
    package_dir={"": "src"},
    packages=setuptools.find_packages(where="src"),
    python_requires=">=3.6",
)
```

---

## LICENSE

- This file provides details on which License is the package distributed under
- Choosing the correct license is often a critical step in open-source software development
- More on licenses in a future chapter of this course

---

## README.md

> - A `README` file typically involves detailed description about the package and also the installation procedure#

Reviewer: ```suggestion
- The `README` file typically gives a description of the package, the code structure, meta information, the installation procedure, ...
```

---

## What is PyPI?

- [PyPI](https://pypi.org/) = **The Python Package Index** is a repository of software developed in the Python community
- PyPI itself is developed on GitHub through another software called [Warehouse](https://github.com/pypa/warehouse)
- PyPI has a informative [public dashboard](https://p.datadoghq.com/sb/7dc8b3250-85dcf667bd) to show its activity
- A major advantage is the active maintenance of PyPI and the packages indexed in it
- Not to be confused with **PyPA** which is Python Packaging Authority, a working group which maintains projects in Python packaging

---

## Example of PyPI package: fenicsprecice

- Having a look at [fenicsprecice](https://pypi.org/project/fenicsprecice/)

---

## Python Enhancement Proposals (PEPs)

- PEP is an evolving design document which provides information regarding new features of Python, new processes and new environments
- PEPs typically involve concise technical information, which also acts as standardizations
- Packaging workflows are also standardized through PEPs. Example:
    - [PEP 427](https://www.python.org/dev/peps/pep-0427/) introduces the built-package format "wheel"

> - It is always a good idea to look for **PEP** conventions before trying out processes

Reviewer 2: ```suggestion
- It is always a good idea to look for [PEP](https://www.python.org/dev/peps/) conventions before trying out processes
```
Why not directly link on the PEPs? This also gives a visual separation/highlight in the slides while making it easy to obtain more information.

---

## File structure for code packaging 1/2

- Most essential component of packaging is file structure. Basic file structure is as follows:

```bash
generic_folder_name/
└── src/

>     └── package_name/

Reviewer 2: In earlier slides it was always called `package-name` with a dash/hyphen instead of the underscore. I just wanted to mention this :wink:

        ├── __init__.py
        └── solver.py
```

> - The file `__init__.py` is required to import the `package_name/` as a package. This file is mostly empty

Reviewer 2: ```suggestion
- The file `__init__.py` is required to import the `package_name/` as a package. This file is normally empty
```
Maybe phrase it like this? I find "mostly" empty confusing. Is it empty or not? I personally have never put any content into a file called `__init__.py`. Are there cases where one would write something in there? If so, it would be nice to give an example (orally or on the slide) where the file has some content.

> - `solver.py` will contain the code developed by the user

Reviewer 2: - Maybe mention here that this can be any other name as well?
- Maybe give an example how one would use the code from `solver.py`?
  
  ```python
  from package_name import solver
  ```

---

## File structure for code packaging 2/2

> - Once the file structuring is complete, the repository will look like:

Reviewer 2: ```suggestion
- Once the file structuring is complete, the repository could look like:
```
Maybe rephrase this or do you already want to point to the exercise?

```bash
generic_folder_name/
├── LICENSE
├── setup.py
├── README.md
├── src/
│   └── package_name/
│       ├── __init__.py

> │       └── solver.py

Reviewer 2: Maybe mention here, later, and/or on the exercise sheet that one should keep tests in mind, but we will talk about this in detail in a later section/chapter of the lecture.

└── tests/
```

<span>
Lets have a detailed look at individual files in this folder structure
<!-- .element: class="fragment" data-fragment-index="1" --></span>

---

> ## Additional options in setup.py 1/2

Reviewer 2: I am a bit confused why this additional information does not come earlier when `setup.py` is introduced. Can I only use these additional options if I use pip or what is the reason for that? Otherwise it might make sense to move this to the beginning of the lecture to keep the part about `setup.py` well connected. Then you could refer here to the previously shown information about the `setup.py` without repeating it.

- [Classifiers](https://pypi.org/classifiers/): additional metadata for the version of the package
- Defined as part of [PEP 303](https://www.python.org/dev/peps/pep-0301/#distutils-trove-classification)
- Example:

```bash
from setuptools import setup

setup(
    ...
    classifiers=[
        "Programming Language :: Python :: <Python-version>",
        "License :: OSI Approved :: <Name of License>",
        "Operating System :: <OS's on which this package runs>",
    ],
    ...
)
```

---

## Additional options in setup.py 2/2

> - The file `README.md` can be passed as a long description in the following way:

Reviewer 2: I would (maybe) move this part to the README slide or to the beginning when talking about the `setup.py`? 

At the moment with this slide and the README slide being so close to each other it feels a lot like repetition. The README slide also mentions that one includes the READMA as longer description.

```bash
from setuptools import setup

with open("README.md", "r", encoding="utf-8") as fh:
    long_description = fh.read()

setup(
    ...
    long_description=long_description,
    long_description_content_type="text/markdown",
    ...
)
```

---

## LICENSE

- This file provides details on which license is the package distributed under
- Choosing the correct license is often a critical step in open-source software development
- More on licenses in a future chapter of this course

---

> ## README.md

Reviewer 2: This headline is rendered as a hyperlink on hedgedoc. Nothing one has to fix, but it is interesting. This also changes the color of the headline to blue.

- A `README.md` file typically gives a description of the package, code structure, meta information, the installation procedure and more ...

> - It is common practice to parse the `README.md` to `setuptools` as a long description

Reviewer 2: Potential repetition as mentioned before.

- Newer versions of `setuptools` also include this file automatically

> - This file can also be written using [reStructuredText](https://docutils.sourceforge.io/rst.html) which is part of [Docutils](https://docutils.sourceforge.io/index.html)

Reviewer 2: ```suggestion
- This file can also be written using [reStructuredText](https://docutils.sourceforge.io/rst.html) which is part of [Docutils](https://docutils.sourceforge.io/index.html). In this case it is usually named `README.rst`
```
I would just make sure that people don't store their reStructureText files wit the markdown ending. ;)

---

## Creating distribution archives

- The project needs to be converted to distribution archives so that pip can install them
- There are two ways to create these archives:
    - Using the package builder [build](https://pypa-build.readthedocs.io/en/stable/index.html) which was introduced in [PEP 517](https://www.python.org/dev/peps/pep-0517/)
    - Using the packaging standard [wheel](https://wheel.readthedocs.io/en/stable/) which was introduced in [PEP 427](https://www.python.org/dev/peps/pep-0427/)
- A source distribution archive file (`package-name-<tags>.tar.gz`) can be generated by running the command `python3 setup.py sdist` in the `package_name/` repository
- A wheel archive file (`package-name-<tags>.whl`) can be generated by running the command `python3 setup.py bdist_wheel` in the `package_name/` repository
- In both cases the files will be generated in a folder `package_name/dist/`

> 

Reviewer 2: I guess you will explain the difference between the two package types?! I am very curious at least to learn about the difference and reasons for different package formats.

- Uploading to TestPyPI can be done using the command: `python3 -m twine upload --repository testpypi dist/*`
- The uploading process looks something like:

```bash
Uploading distributions to https://test.pypi.org/legacy/
Enter your username: [your username]
Enter your password: [your password]
Uploading package_name-0.0.1-py3-none-any.whl
100%|█████████████████████| 4.65k/4.65k [00:01<00:00, 2.88kB/s]
Uploading package_name-0.0.1.tar.gz
100%|█████████████████████| 4.25k/4.25k [00:01<00:00, 3.05kB/s]
```

---

## Installing the uploaded package

- The last step in the packaging pipeline is a sanity check done by installing the package that has been uploaded
- The installation can be done by:

```bash
pip install --user --index-url https://test.pypi.org/simple/ <package-name>
```

---

## Conda

- [Conda](https://docs.conda.io/en/latest/) is a package and environment management system which supports multiple languages
- Conda provides a fast option to setup an isolated environment on your local system
- Conda is configured to work with [Anaconda installers and packages](https://repo.anaconda.com/)

> - Conda is often the preferred way to run Python packages on Windows and MacOS. See example in [FEniCS on Anaconda section](https://fenicsproject.org/download/)

Reviewer 2: Is it not popular on Linux? I have only used it on Linux so far. :smile:

Author: Hence the *often*, meaning that excluding Linux it is also preferred in ....

---

## Difference between Conda and pip

| Feature | Conda | pip |
| Multi-language dependency | Yes | No |
| Package installation | Anaconda installer / download binaries | build / wheels |
| Package availability | ~8000 | ~336,000 |

> | Virtual environment functionality | In-built environment management | No, but support for `virtualenv` |

Author: Table has been repaired. I will briefly talk about `virtualenv` while talking about this table.

> ---

Reviewer 2: Do you show something practical about conda here? Homepage, how to create an environment or similar? Otherwise I am not sure whether one learns much from this section.

Author: No demo for Conda due to time constraints. Conda is more of an FYI in this context.

## Important takeaways

> Packaging involves two central work packages:

Reviewer 2: ```suggestion
Packaging of Python code involves two central work packages:
```

> - Creation of a standardized folder structure to convert a raw code into a project

Reviewer 2: ```suggestion
- Creation of a standardized folder structure to convert raw Python code into a project
```
